### DISCUSSION TITLE
Improve developer experience by allowing code errors surface

### DISCUSSION THREAD

**Discussion by ledestin** [forem/forem#15315] [Feature Requests]
**Is your feature request related to a problem? Please describe.**

While working on request specs of comments controller, I found out that exceptions wouldn't surface, when running request specs. The culprit turned out to be `resque StandardError`:
https://github.com/thepracticaldev/dev.to/blob/master/app/controllers/comments_controller.rb#L114

The problem here is two-fold:
1. To get to the error, I need to analyse response payload, and I don't get a backtrace. Given that the project is set up to surface exceptions via `config.action_dispatch.show_exceptions = false`, this practice breaks the default behaviour, which worsens developer experience.
2. `NameError` and `NoMethodError` are subclasses of `StandardError`:
```
ruby -e 'wtf'
Traceback (most recent call last):
-e:1:in `<main>': undefined local variable or method `wtf' for main:Object (NameError)
```
Thus, I can write nonsense, and I will not know, unless I take care to examine response payload. I hit this problem while fixing a bug, and I lost time on it.

`rescue StandardError` is all over the place.

**Describe the solution you'd like**

I'd like programming errors surface as exceptions with a proper backtrace, in test environment. I suggest rescuing `RuntimeError`, instead of `StandardError`.

**Describe alternatives you've considered**

For request specs, it's examining payload, but for models, there's no solution like that.

**Comment by Ridhwana** [comment]
Thanks for the issue! We'll take your request into consideration and follow up if we decide to tackle this issue.

To our amazing contributors: issues labeled `type: bug` are always up for grabs, but for feature requests, please wait until we add a `ready for dev` before starting to work on it.

To claim an issue to work on, please leave a comment. If you've claimed the issue and need help, please ping @thepracticaldev/oss and we will follow up within 3 business days.

For full info on how to contribute, please check out our [contributors guide](https://github.com/thepracticaldev/dev.to/blob/master/CONTRIBUTING.md).

**Comment by rhymes** [comment]
I think this request also sits in a larger conversation about using custom exceptions as much as possible instead of `StandardError` which can have the byproduct of silencing meaningful ones.

I'm not sure what the perfect approach is but it's worth a discussion

**Comment by ledestin** [comment]
What was the original intent of rescuing `StandardError` and not `RuntimeError`?

**Comment by rhymes** [comment]
@ledestin hard to say, If I had to guess, it's because all liquid tags (that can be rendered in comments) raise StandardError (which we plan to fix eventually) and not a custom exception.

Thus one needs to handle that at the controller level to be able to return rendering errors.

BTW personally I'm -1 on using `RuntimeError` as much as possible, +1 on using custom and meaningful exceptions

**Comment by zakhalex** [comment]
@ledestin @rhymes 

While not an active contributor to this project (so take my opinion with a grain of salt) I'd personally vote +1 on @ledestin 's original proposal to use RuntimeError with proper exception propagation, as custom exceptions add little value, but have a management/support overhead.

**Comment by rhymes** [comment]
@zakhalex it's just that`RuntimeError` is described as `A generic error class raised when an invalid operation is attempted.` and I still believe we shouldn't lean on generic exceptions :D 

Anyway, that doesn't mean the argument is settled!

**Comment by ledestin** [comment]
@rhymes it's possible to `ApplicationError < RuntimeError` and `CustomError < ApplicationError`, thus both having custom exceptions and not having to deal with specifying them in `rescue`.

How about scraping `rescue StandardException` everywhere and adding `rescue_from StandardException unless Rails.env.test?`, making `ApplicationController` the handler of those exceptions. It's a temporary solution that can improve developer experience.

**Comment by rhymes** [comment]
@ledestin I think we should aim for a lasting solution that can be introduced bit by bit. We already have a few custom exceptions around, both in `app/errors` and sprinkled in `app/services`. I would start by analyzing what's needed, what has to go and what has to be changed. It's not an easy change :-(

**Comment by ledestin** [comment]
@rhymes the problem with a solution that's slow to be implemented, is that development speed would be impeded by concealed exceptions. I had to comment out `rescue StandardError`, when I working on that bugfix. That is, after I spent time figuring out what was happening.

What are downsides of my solution? It's not like custom exceptions can't be gradually introduced afterwards.

**Comment by rhymes** [comment]
The code shouldn't behave differently in testing mode as much as possible :) 

There are 41 instances of `rescue StandardError`, adding "unless Rails.env.test?" to all of them would mean that we would break a lot of tests, then we'd need to fix those. The result would be a PR that doesn't really fix the problem

**Comment by ledestin** [comment]
@rhymes I thought the problem was that errors were swallowed. If they aren't swallowed in the test env, it's still an improvement.

How about just replacing `StandardError` with `ApplicationError < RuntimeError`? That would retain the uniqueness that `StandardError` has now, but errors won't be swallowed anymore? Later on, `ApplicationError` can be changed to custom errors one by one.

**Comment by rhymes** [comment]
I still think we should change one by one in context and fix the underlying issue of having an exception that's too generic instead. But I welcome a PR so we the team can look over it

**Comment by ledestin** [comment]
@rhymes I've tried, and, there were too many various issues, like `NoMethodError`, all over the place. It's not an easy fix.

Would you be happy with introduction of `ApplicationError`? If I encounter `StandardError` while working on something, I'd make a PR fixing it and mention this issue, and use `ApplicationError` as a replacement for `StandardError`. What do you think?

**Comment by rhymes** [comment]
I think the strategy it's okay but I'd rather have a POC with a small change that the rest of the team can weigh in as it shouldn't be the decision of a single developer.

AI agree that we need to improve error handling, I'm not sure we're having the same development experience in terms of problems created by the present error handling.

This might make it easier to have a collective discussion

**Comment by citizen428** [comment]
> Given that the project is set up to surface exceptions via `config.action_dispatch.show_exceptions = false`, this practice breaks the default behaviour

I don't see it this way. It's the default behavior for **unhandled** exceptions, which defines how Rails handles them. However, we are handling them. The handling may certainly be suboptimal which is the whole point of this discussion, but the value of this setting is IMHO not a sufficient justification for the change by itself.

> Thus, I can write nonsense, and I will not know

The specific code path you linked does log the error, did you check `log/test.log`?

None of this is to say that I don't think there's a problem with how we currently handle errors, just that I'm not entirely convinced by the reasons outlined above.

Having an `ApplicationError` that defaults to `RuntimeError` could work. It's somewhat easier since people don't have to pick between `StandardError` and `RuntimeError` by themselves, but we need to document it somewhere (and/or add a simple generator). Also: both `raise` and `fail` raise `RuntimeError`s by default when no error class is specified, so in one-off cases, we may not even need a custom error class, just a good enough message.

**Comment by ledestin** [comment]
>>Given that the project is set up to surface exceptions via config.action_dispatch.show_exceptions = false, this practice breaks the default behaviour

>I don't see it this way. It's the default behavior for unhandled exceptions, which defines how Rails handles them. However, we are handling them. The handling may certainly be suboptimal which is the whole point of this discussion, but the value of this setting is IMHO not a sufficient justification for the change by itself.

@citizen428 Well, it's up to the team on how you want to handle it, of course.

>> Thus, I can write nonsense, and I will not know

> The specific code path you linked does log the error, did you check log/test.log?

> None of this is to say that I don't think there's a problem with how we currently handle errors, just that I'm not entirely convinced by the reasons outlined above.

IIRC, I've seen it there (without a stacktrace), but that's not the workflow I'd like to follow. Having errors bubble is faster and easier to deal with for me.

> Having an ApplicationError that defaults to RuntimeError could work. It's somewhat easier since people don't have to pick between StandardError and RuntimeError by themselves, but we need to document it somewhere (and/or add a simple generator). Also: both raise and fail raise RuntimeErrors by default when no error class is specified, so in one-off cases, we may not even need a custom error class, just a good enough message.

Do you think that `StandardError` should be raised/rescued at all?

Documenting it is a great idea.

**Comment by citizen428** [comment]
> Do you think that StandardError should be raised/rescued at all?

With the classifier "at all", 100% yes, as there'll always be circumstances where a specific combination makes more sense than the other. 

The Ruby documentation has the following to say on the matter:

> Programs may make subclasses of `Exception`, typically of `StandardError` or `RuntimeError`, to provide custom classes and add additional information.

The defaults for unqualified `raise` and `rescue` are `RuntimeError` and `StandardError` respectively. Based on that having `ApplicationError` (or any more specific error class) inherit from `RuntimeError` seems absolutely sensible. Whether or not we want to narrow the focus for `rescue` is a more subtle issue: there are certainly cases where I don't care what the error is, I just want to get on with things (e.g. big batch jobs, where I don't care if a handful of records fail, just log the error and be done with it). That said, it's probably better to be more specific in most cases and explicitly be more forgiving (`StandardError`) in the few cases where it makes sense.

Based on all of that I suggest the following:

1. Have an `ApplicationError` class which inherits from `RuntimeError` as the base class for our custom exceptions.
2. Custom exceptions classes should most likely inherit from that, though under certain circumstances may want to chose a different base class.
3. Our default for `rescue` should switch from `StandardError` to `RuntimeError`, not `ApplicationError`. If we switch to `ApplicationError` all other `RuntimeErrors` will not be rescued anymore, which may be more than we're looking for. 

I think this is a nice middle-ground between accidentally swallowing `NoMethodError` and similar exceptions and only handling our own custom exceptions.

Sorry for the somewhat long-winded answer, but this is a potentially far-reaching change so I want to think this through properly. 😃 

> Documenting it is a great idea.

Whatever we decide on needs to survive the discussions in this issue 😃

**Comment by rhymes** [comment]
I agree with @citizen428, I would add that we should probably do this incrementally, not a big on-off switch. We can raise a custom exception descended by `ApplicationError` in one liquid tag, test that in production and then start using it. 

Agreed that our exceptions structure should be, eventually, fully documented in a specific page on docs.dev.to - we can start with whatever comes out from this discussion.

**Comment by citizen428** [comment]
@rhymes and I had a Slack conversation about this. I summarized my ideas one more time:

1. Custom exceptions often make sense and are preferred.
2. IMHO it’s better if they inherit from `RuntimeError` rather than `StandardError`. The default for `raise` suggests that too. Exceptions (no pun) would be errors raised during init or similar, as that’s not strictly speaking runtime.
3. Our "broader" rescue statements (not just one specific exception class) should not potentially not catch `StandardError`, but `RuntimeError`. If we have a `NoMethodError`, I want to see that in Honeycomb or somewhere, not caught by an over-eager `rescue` and logged.

tl;dr: I care less about what we `raise` but more about what we `rescue`, which I believe was @ledestin's original point.

**Comment by ledestin** [comment]
@rhymes I've updated https://github.com/forem/forem/pull/8765, and now it just introduces `ApplicationError` and raises it in one of the liquid tags.

But, we can't switch `rescue StandardError` to `RuntimeError` in controllers after changing just that. Given the changes to error handling, errors end up in Honeybadger now. Those should be fixed one by one. After those errors are fixed (stuff like `NoMethodError`), eventually, we can switch to rescuing `RuntimeError`. Rescuing both `StandardError` and `RuntimeError` doesn't make sense as `RuntimeError < StandardError`, and `RuntimeError` is rescued currently.

**Comment by rhymes** [comment]
@ledestin why not rescueing `ApplicationError` or even better a `LiquidTagError < ApplicationError < RuntimeError` ?
